supports openharmony for v3.17.2

zhaoyang57 opened the pull request.

zhaoyang57 committed:
supports openharmony for v3.17.2

zhaoyang57 committed:
supports openharmony for v3.17.2

qiuguohua commented on extensions/Particle3D/PU/CCPUMaterialManager.cpp:
```diff
@@ -159,7 +164,17 @@ bool PUMaterialCache::loadMaterialsFromSearchPaths( const std::string &fileFolde
+    for (auto fileName : files) {
+        if (FileUtils::getInstance()->getFileExtension(fileName) == ".material")
+        {
+            std::string fullpath = std::string(fileName);
```
The type of fileName is std::string?

1-newbie replied:
yes，string，we modified it，will submit

dumganhar commented on cocos/audio/AudioEngine.cpp:
```diff
@@ -152,6 +154,9 @@ class AudioEngine::AudioEngineThreadPool
 void AudioEngine::end()
 {
+#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
+    stopAll();
```
Why does OHOS need to invoke `stopAll` in `end` but other platforms don't?

1-newbie replied:
recovered

qiuguohua commented on cmake/Modules/CocosBuildSet.cmake:
```diff
@@ -24,8 +24,8 @@ message(STATUS "ENGINE_BINARY_PATH:" ${ENGINE_BINARY_PATH})
-option(BUILD_LUA_LIBS "Build lua libraries" OFF)
-option(BUILD_JS_LIBS "Build js libraries" OFF)
+option(BUILD_LUA_LIBS "Build lua libraries" ON)
+option(BUILD_JS_LIBS "Build js libraries" ON)
```
Other platforms turn it on by default? Why not use the configuration?

1-newbie replied:
We need to use luatests and jstests to test the functionality of the engine, but the default value is off, making each modification more cumbersome. Therefore, we choose to keep this switch on. The next update of PR will close it

qiuguohua commented on cocos/platform/ohos/CCApplication-ohos.cpp:
```diff
@@ -0,0 +1,114 @@
+#if CC_TARGET_PLATFORM == CC_PLATFORM_OHOS
```
Why do we need this macro, this file should be specific to openharmony?

1-newbie replied:
CCApplication-android.h has the similar macro, we are just trying to maintain consistency in style

qiuguohua commented on cocos/platform/ohos/CCApplication-ohos.h:
```diff
@@ -0,0 +1,89 @@
+#define __CC_APPLICATION_OHOS_H__
+
+#include "platform/CCPlatformConfig.h"
+#if CC_TARGET_PLATFORM == CC_PLATFORM_OHOS
```
Why do we need this macro, this file should be specific to openharmony?

1-newbie replied:
CCApplication-android.h has the similar macro, we are just trying to maintain consistency in style

minggo commented on cocos/audio/AudioEngine.cpp:
```diff
@@ -378,6 +404,22 @@ void AudioEngine::uncache(const std::string &filePath)
     if (audioIDsIter != _audioPathIDMap.end())
     {
+	#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
+        auto lst =  _audioPathIDMap[filePath];
```
- audioIDsIter contains the result, so can just use it
- and if still need to use it like this, i think it is better to use const reference to improve performance

1-newbie replied:
recovered

minggo commented on cocos/audio/AudioEngine.cpp:
```diff
@@ -378,6 +404,22 @@ void AudioEngine::uncache(const std::string &filePath)
     {
+	#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
+        auto lst =  _audioPathIDMap[filePath];
+        for (auto it = lst.begin() ; it != lst.end(); ++it) {
```
How about use `for in` loop?

1-newbie replied:
recovered

minggo commented on cocos/audio/include/AudioEngine.h:
```diff
@@ -175,6 +175,11 @@ class EXPORT_DLL AudioEngine
+    #if CC_TARGET_PLATFORM == CC_PLATFORM_OHOS
+    static void onEnterBackground();
+    static void onEnterForeground();
+    #endif
```
Why need these functions?

1-newbie replied:
will delete

dumganhar commented on cocos/base/ccConfig.h:
```diff
@@ -162,7 +162,11 @@ THE SOFTWARE.
 #ifndef CC_TEXTURE_ATLAS_USE_VAO
-#define CC_TEXTURE_ATLAS_USE_VAO 1
+#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
+    #define CC_TEXTURE_ATLAS_USE_VAO 0 // TBD ohos doesn't support VAO
```
VAO is a standard of OpenGLES 3.x. If ohos supports OpenGLES 3.x, it should be able to support VAO.

1-newbie replied:
yes, ohos supports GLES3, but not support GLES2 yet. if we set 1 here, it will allow to execute the follow code. Therefore, we turned off support for VAO and blocked the following code through macros
![image](https://github.com/cocos2d/cocos2d-x/assets/57482234/b4561561-8beb-4d16-b881-1a4ad98979fa)

dumganhar replied:
So you mean that there is not a `glMapBufferOES` extension in GLES2 implementation on OH platform?

dumganhar replied:
<img width="963" alt="企业微信截图_02435cc0-c72c-4384-bca4-e2727e3d18df" src="https://github.com/cocos2d/cocos2d-x/assets/493372/d8beb1df-1393-478e-ab82-6efae74ec59d">

1-newbie replied:
yes, no gles2 on ohos now. this header file is pre-added.
in the cmakelist, no link to GLES2 now

dumganhar commented on cocos/platform/CCApplicationProtocol.h:
```diff
@@ -58,7 +58,8 @@ class CC_DLL ApplicationProtocol
         OS_WINRT,       /**< Windows Runtime Applications */
-        OS_WP8          /**< Windows Phone 8 Applications */
+        OS_WP8,         /**< Windows Phone 8 Applications */
+        OS_HARMONY      /**< HARMONY */
```
```suggestion
        OS_OPENHARMONY      /**< OPENHARMONY */
```

minggo commented on cocos/renderer/CCRenderer.cpp:
```diff
@@ -793,30 +793,33 @@ void Renderer::drawBatchedTriangles()
+       
+       glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, _buffersVBO[1]);
+       glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(_indices[0]) * _filledIndex, _indices, GL_STATIC_DRAW);
+#endif
```
It seems the content is the same, why has so many modifications?

1-newbie replied:
we just add macro to avoid compile error for not support GLES2, no idea to why has so many modifications

1-newbie replied:
we just add macro to avoid compile error for not support GLES2. It seems the indent caused so many modifications, we will restore it

minggo commented on cocos/renderer/CCTextureAtlas.cpp:
```diff
@@ -618,45 +618,47 @@ void TextureAtlas::drawNumberOfQuads(ssize_t numberOfQuads, ssize_t start)
+       glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+#endif
 #endif
-
```
The same as above, did you modify the ident?

1-newbie replied:
ok, will recover

dumganhar commented on cocos/platform/ohos/CCGL-ohos.h:
```diff
@@ -0,0 +1,43 @@
+#ifndef __CCGL_H__
+#define __CCGL_H__
```
```suggestion
#ifndef __CCGL_OHOS_H__
#define __CCGL_OHOS_H__
```

minggo commented on cocos/scripting/lua-bindings/script/cocos2d/Cocos2dConstants.lua:
```diff
@@ -188,6 +188,7 @@ cc.PLATFORM_OS_EMSCRIPTEN = 8
 cc.PLATFORM_OS_TIZEN   = 9
 cc.PLATFORM_OS_WINRT   = 10
 cc.PLATFORM_OS_WP8     = 11
+cc.PLATFORM_OS_OHOS     = 12
```
It is better to align `=`.

1-newbie replied:
ok

dumganhar commented on cocos/platform/ohos/CCLogOhos.h:
```diff
@@ -0,0 +1,121 @@
+#ifndef DISP_COMMON_H
```
What's DISP_COMMON_H ?
Should it be `#ifndef __CC_LOG_OHOS_H__` ?

1-newbie replied:
ok, modified, will submit

dumganhar commented on cocos/platform/ohos/CCPlatformDefine-ohos.h:
```diff
@@ -0,0 +1,37 @@
+#ifndef __CCPLATFORMDEFINE_H__
```
```suggestion
#ifndef __CCPLATFORMDEFINE_OHOS_H__
```

dumganhar commented on cocos/platform/ohos/CCStdC-ohos.h:
```diff
@@ -0,0 +1,23 @@
+#ifndef __CC_STD_C_H__
```
```suggestion
#ifndef __CC_STD_C_OHOS_H__
```

dumganhar commented on cocos/platform/ohos/CCTextBitmap.cpp:
```diff
@@ -0,0 +1,204 @@
+CCTextBitmap::~CCTextBitmap()
+{
+  OH_Drawing_CanvasDestroy(_canvas);
+//  OH_Drawing_BitmapDestroy(_bitmap);
```
Doesn't `_bitmap` need to be released?

1-newbie replied:
the destructor function of CCTextBitmap seems never called. we fixed this bug

dumganhar commented on cocos/scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.cpp:
```diff
@@ -1,5 +1,5 @@
 #include "scripting/js-bindings/auto/jsb_cocos2dx_audioengine_auto.hpp"
-#if (CC_TARGET_PLATFORM == CC_PLATFORM_WINRT || CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID || CC_TARGET_PLATFORM == CC_PLATFORM_IOS || CC_TARGET_PLATFORM == CC_PLATFORM_MAC || CC_TARGET_PLATFORM == CC_PLATFORM_WIN32)
+#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS || CC_TARGET_PLATFORM == CC_PLATFORM_WINRT || CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID || CC_TARGET_PLATFORM == CC_PLATFORM_IOS || CC_TARGET_PLATFORM == CC_PLATFORM_MAC || CC_TARGET_PLATFORM == CC_PLATFORM_WIN32)
```
https://github.com/cocos2d/cocos2d-x/blob/v3/tools/tojs/cocos2dx_audioengine.ini#L10C21-L10C26

This file should also be updated.

dumganhar replied:
The same as other `_auto` suffix files.

1-newbie replied:
![image](https://github.com/cocos2d/cocos2d-x/assets/57482234/ebbcb9e6-33f2-46e3-a79b-5c1e684d1f11)

Are you saying we need modify all the files under tojs and tolua like this?

1-newbie replied:
We haven't used this tool, so we haven't focused on this part before.

dumganhar replied:
Yes. `*_auto.cpp` files are generated by `tolua/tojs` tools automatically.

1-newbie replied:
ok

dumganhar commented on cocos/scripting/lua-bindings/auto/lua_cocos2dx_audioengine_auto.cpp:
```diff
@@ -1,5 +1,5 @@
 #include "scripting/lua-bindings/auto/lua_cocos2dx_audioengine_auto.hpp"
-#if CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID || CC_TARGET_PLATFORM == CC_PLATFORM_IOS || CC_TARGET_PLATFORM == CC_PLATFORM_MAC || CC_TARGET_PLATFORM == CC_PLATFORM_WIN32 || CC_TARGET_PLATFORM == CC_PLATFORM_TIZEN
+#if CC_TARGET_PLATFORM == CC_PLATFORM_OHOS || CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID || CC_TARGET_PLATFORM == CC_PLATFORM_IOS || CC_TARGET_PLATFORM == CC_PLATFORM_MAC || CC_TARGET_PLATFORM == CC_PLATFORM_WIN32 || CC_TARGET_PLATFORM == CC_PLATFORM_TIZEN
```
Should also modify https://github.com/cocos2d/cocos2d-x/blob/v3/tools/tolua/cocos2dx_audioengine.ini#L10

dumganhar commented on cocos/scripting/lua-bindings/script/framework/device.lua:
```diff
@@ -58,6 +58,8 @@ elseif target == cc.PLATFORM_OS_WINRT then
 elseif target == cc.PLATFORM_OS_WP8 then
     device.platform = "wp8"
+elseif target == cc.PLATFORM_OS_OHOS then
+    device.platform = "harmony"
```
```suggestion
    device.platform = "openharmony"
```

dumganhar commented on cocos/ui/CMakeLists.txt:
```diff
@@ -70,6 +70,20 @@ elseif(ANDROID)
         # it's special for android, not a common file
         ui/UIWebView.cpp
         )
+ELSEIF(OHOS)
```
```suggestion
elseif(OHOS)
```

dumganhar commented on cocos/ui/UIVideoPlayer-ohos.h:
```diff
@@ -0,0 +1,5 @@
+#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
```
Add `#pragma once`

1-newbie replied:
added

dumganhar commented:
Why add `tests/lua-empty-test/res/farm.png` file?
There is already a `farm.jpg` in that folder. Doesn't openharmony support `jpeg decoding`?

minggo commented:
About `doc/game-porting-guide`:
- it is better to named `docs/openharmony/game-porting-guide`
- please don't use chinese in folder name, such as `OpenHarmony系统架构.png`

minggo commented on extensions/assets-manager/AssetsManager.cpp:
```diff
@@ -197,7 +197,6 @@ std::string AssetsManager::keyOfDownloadedVersion() const
 bool AssetsManager::checkUpdate()
 {
     if (_versionFileUrl.size() == 0 || _isDownloading) return false;
-    
```
Having a space line makes codes more clear. So i don't think need to remove it.

1-newbie replied:
recovered

minggo commented on tests/cpp-tests/CMakeLists.txt:
```diff
@@ -242,6 +249,8 @@ list(APPEND GAME_SOURCE
      Classes/CurlTest/CurlTest.cpp
      Classes/CurrentLanguageTest/CurrentLanguageTest.cpp
+     Classes/SDKTest/SDKTest.cpp
+     Classes/PaySDKTest/PaySDKTest.cpp
```
Can these tests work on all platforms?

minggo replied:
And PaySDK is not part or engine, so it is not suitable to test it here.

1-newbie replied:
ok, deleted

minggo commented on tests/cpp-tests/Classes/PaySDKTest/PaySDKTest.cpp:
```diff
@@ -0,0 +1,197 @@
+  addChild(label, 0);
+  label->setPosition(VisibleRect::center().x, VisibleRect::top().y - 50);
+
+    //检查支付环境
```
Please don't use chinese in comment.

1-newbie replied:
deleted

minggo commented on tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/WebSocketDelayTest.cpp:
```diff
@@ -100,7 +100,7 @@ void WebSocketDelayTest::startTestCallback(Ref* sender)
     protocols.push_back("myprotocol_1");
     protocols.push_back("myprotocol_2");
-    if (!_wsiSendText->init(*this, "wss://echo.websocket.org", &protocols, "cacert.pem"))
+    if (!_wsiSendText->init(*this, "ws://echo.websocket.events", &protocols, "cacert.pem"))
```
Why modify it?

1-newbie replied:
I remember the orignal one doesn't work

minggo commented on tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/WebSocketTest.cpp:
```diff
@@ -137,7 +137,8 @@ void WebSocketTest::startTestCallback(Ref* sender)
     protocols.push_back("myprotocol_2");
-    if (!_wsiSendText->init(*this, "wss://echo.websocket.org", &protocols, "cacert.pem"))
+    // if (!_wsiSendText->init(*this, "wss://echo.websocket.events", &protocols, "cacert.pem"))
+    if (!_wsiSendText->init(*this, "ws://echo.websocket.events"))
```
The same as above.

1-newbie replied:
the orignal one doesn't work

minggo commented:
Why the files in `tests/cpp-tests/proj.ohos/dependencies`, such as `hvigor-3.0.4-s.tgz`,  needed?

yyt233 committed:
移除sdktest/paysdktest

yyt233 committed:
移除sdktest/paysdktest

yyt233 committed:
Clarify that the type of fileName is string, remove redundant code

yyt233 committed:
Clarify that the type of fileName is string, remove redundant code

yyt233 committed:
Restore to initial value

yyt233 committed:
Restore to initial value

1-newbie committed:
Update cocos/platform/CCApplicationProtocol.h

Co-authored-by: James Chen <dumganhar@gmail.com>

yyt233 committed:
1. delete the onEnterBackground in the base class
2. some new configurations

1-newbie committed:
Update cocos/platform/ohos/CCGL-ohos.h

Co-authored-by: James Chen <dumganhar@gmail.com>

yyt233 committed:
recover the indent

1-newbie committed:
Update cocos/platform/ohos/CCPlatformDefine-ohos.h

Co-authored-by: James Chen <dumganhar@gmail.com>

1-newbie committed:
Update cocos/platform/ohos/CCStdC-ohos.h

Co-authored-by: James Chen <dumganhar@gmail.com>

1-newbie committed:
Update cocos/scripting/lua-bindings/script/framework/device.lua

Co-authored-by: James Chen <dumganhar@gmail.com>

1-newbie committed:
Update cocos/ui/CMakeLists.txt

Co-authored-by: James Chen <dumganhar@gmail.com>

yyt233 committed:
1. macro add OHOS
2. CCTextBitmap need delete after use
3. UIVideoPlayer-ohos.h add #pragma once
4. all docs under OpenHarmony and change the filename from chinese to english
5. recover the line in AssetsManager.cpp

yyt233 committed:
recover some modifications in AudioEngine.cpp

yyt233 committed:
add OHOS to each file under tojs/tolua

yyt233 committed:
recover unnecessary modifications

yyt233 committed:
recover unnecessary modifications

qiuguohua commented on cocos/base/CCController-ohos.cpp:
```diff
@@ -0,0 +1,164 @@
+
+#include "base/CCController.h"
+
+#if (CC_TARGET_PLATFORM == CC_PLATFORM_OHOS)
```
This is a file unique to ohos and should not require the addition of macro definitions?

1-newbie replied:
just keep the code style, cause android has the same macro

qiuguohua replied:
Okay

qiuguohua commented on cocos/platform/CCCommon.h:
```diff
@@ -37,6 +37,9 @@ NS_CC_BEGIN
  */
 
+/// The max length of CCLog message.
+static const int kMaxLogLen = 16*1024;
```
What is this definition used for?

1-newbie replied:
deleted

qiuguohua commented on cocos/platform/ohos/CCApplication-ohos.cpp:
```diff
@@ -0,0 +1,114 @@
+int Application::run()
+{
+    // Initialize instance and cocos2d.
+    if (! applicationDidFinishLaunching())
```
extra space

1-newbie replied:
ok, removed

qiuguohua commented on cocos/platform/ohos/CCDevice-ohos.cpp:
```diff
@@ -0,0 +1,112 @@
+        bitmapDC.m_nWidth = cCtextBitmap->GetWidth();
+        bitmapDC.m_nHeight = cCtextBitmap->GetHeight();
+        long size = bitmapDC.m_nWidth * bitmapDC.m_nHeight * 4;
+        bitmapDC.m_pData = new unsigned char[size];
```
sharedBitmapDC is a static variable and will be freed when the program exits, if it is reused, there will be memory leaks

1-newbie replied:
the m_pData will be returned as part of Data in the getTextureDataForText function and might be freed at somewhere.
in the similar function of android, it also malloc new memory every time.
final, we use Profile(IDE tools) to detected the memory usage, no noticeable memory leaks issues found

1-newbie commented:
> Why the files in `tests/cpp-tests/proj.ohos/dependencies`, such as `hvigor-3.0.4-s.tgz`, needed?

cause the deveco still in beta, the developer can't download hvigor from the remote repo, so we need use the local file. After the hvigor can download, we will remove these files.

yyt233 committed:
dekete extra space

yyt233 committed:
necessary definitions

yyt233 committed:
申请空间从new改成malloc

dumganhar commented on cocos/platform/ohos/CCDevice-ohos.cpp:
```diff
@@ -0,0 +1,112 @@
+        bitmapDC.m_nWidth = cCtextBitmap->GetWidth();
+        bitmapDC.m_nHeight = cCtextBitmap->GetHeight();
+        long size = bitmapDC.m_nWidth * bitmapDC.m_nHeight * 4;
+        bitmapDC.m_pData = (unsigned char*)malloc(sizeof(unsigned char) * size);
```
m_pData is released by `delete [] m_pData;` in `~BitmapDC`. Why change to `malloc` ?

yyt233 committed:
keep same with android, release the m_pData every time after render
